### DISCUSSION TITLE
How to import production (larger than toy size) data

### DISCUSSION THREAD

**Discussion by sdgathman** [adempiere/adempiere#4120] [Q&A] [answered]
The import process use an in memory file uploaded by the client.  The largest such csv file in the case I am working on is 100M.  That is easily handled in the JVM after increasing max memory option.  However, the web client is unable to do such large uploads through http.

My immediate workaround is to patch the import process to use a stream from a fixed local (to Adempiere server) file.  I can rsync the data being worked on.  

As a simple more general solution, it should work like the EDI import, where you can upload an edi document, it fetches from an ftp url (and eventually http), or you enter the full path to a copy of the document on the Adempiere server.

**Comment by e-Evolution** [comment]
@sdgathman needs to compromise ZK configuration https://forum.zkoss.org/question/39230/how-to-set-the-max-upload-size-for-fileupload/

Or use the Swing interface 

kind regards, 
Victor

**Comment by yamelsenih** [comment, marked as answer]
Hello @sdgathman the main problem is that the file should be split and process each file separately.

This is a nice change for import loader.

Other problem is that if you have many records to import then the import process is very very slowly because this use the Persistence Object and all business logic of ADempiere, I personally think that a way for resolve it is process by batch instead all but is a big work change all import process.

Best regards

**Reply by sdgathman** [reply]
So there are 2.4 million records to import - 100Mbyte.  Do you think 10 batches of 240K records / 10M is reasonable?
100 batches of 24K records / 1M ?

**Reply by yamelsenih** [reply]
Maybe you can test it. Other way is improve the process to process each batch but is a hard work also

**Comment by e-Evolution** [comment]
@sdgathman Hi , if you want to improve the performance and you can use multiple cpus, you can do a refactory in the import of products to work in parallel as a result of the order.

https://github.com/adempiere/adempiere/blob/develop/base/src/org/compiere/process/ImportOrder.java#L228

 kind regards
Victor

**Reply by sdgathman** [reply]
This is the initial import of the product list - not a huge number of items on a sales/purchase order.

**Comment by sdgathman** [comment]
Can I just use a postgresql queries to populate the I_Product table?  Isn't that a staging table that is then transferred to M_Product and others?  I guess there is some verification logic for I_Product as well.

**Reply by yamelsenih** [reply]
Off course, just you should care something:
- Product sequence ID should be checked after it, I use a function to generate it `nextidfunc`
- The UOM is mandatory: This field can't be changed after the first inventory record
- The Product Category is mandatory: You should define this before any transaction
- The Value: Product Code
- The Name: Product Name
- The Tax Category
- Product Type: Define if is a product as item, resource or service

The rest of fields just keep the default value (if only is product).

Note that exists some tables that are need for AD functionality like Cost, Accounting and Translation:

- M_Product
- M_Product_Trl: Can be re-generated with process for generate translation by client or a `INSERT SELECT` from product with the same name
- M_Product_Acct: Can be re-generated using the copy account process from **Product Category** window
- M_Product_Costing: Currently have half support
- M_Cost: Can be re-generated using the process **Generate Cost transaction** => Very Slowly because is a product of Account Schema, Costing Level, Warehouse and product, however, the first generation is more fast. also you can wait for first transaction like PO and this will be generated, other option is use the process **Create Cost Record**
- AD_TreeNodePR: Can be re-generated with the process **Tree Maintenance** inside **Tree** window

I test it in Garden World
![Screenshot from 2023-08-01 11-57-39](https://github.com/adempiere/adempiere/assets/2333092/6fc57d7a-2a02-49c1-bfbf-c5d01bdd9997)

Here the result

```SQL
-- Aug 1, 2023, 11:58:23 AM VET
-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
INSERT INTO M_Product (AD_Client_ID,AD_Org_ID,CopyFrom,Created,CreatedBy,C_TaxCategory_ID,C_UOM_ID,Discontinued,GuaranteeDays,GuaranteeDaysMin,IsActive,IsBOM,IsDropShip,IsExcludeAutoDelivery,IsInvoicePrintDetails,IsKanban,IsManufactured,IsPhantom,IsPickListPrintDetails,IsPurchased,IsSelfService,IsSold,IsStocked,IsSummary,IsToFormule,IsVerified,IsWebStoreFeatured,LowLevel,M_Locator_ID,M_Product_Category_ID,M_Product_ID,Name,Processing,ProductType,ShelfDepth,ShelfHeight,ShelfWidth,UnitsPerPack,UnitsPerPallet,Updated,UpdatedBy,UUID,Value,Volume,Weight) VALUES (11,0,'N',TO_DATE('2023-08-01 11:58:22','YYYY-MM-DD HH24:MI:SS'),100,107,100,'N',0,0,'Y','N','N','N','N','N','N','N','N','Y','Y','Y','Y','N','N','N','N',0,50006,105,1000000,'Test','N','I',0,0,0,0,0,TO_DATE('2023-08-01 11:58:22','YYYY-MM-DD HH24:MI:SS'),100,'9b337997-c280-4da2-9f6e-9123eec0cb03','01',0,0)
;

-- Aug 1, 2023, 11:58:23 AM VET
-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
INSERT INTO M_Product_Trl (AD_Client_ID,AD_Language,AD_Org_ID,Created,CreatedBy,IsActive,IsTranslated,M_Product_ID,Name,Updated,UpdatedBy,UUID) VALUES (11,'es_MX',0,TO_DATE('2023-08-01 11:58:23','YYYY-MM-DD HH24:MI:SS'),100,'Y','N',1000000,'Test',TO_DATE('2023-08-01 11:58:23','YYYY-MM-DD HH24:MI:SS'),100,'43e4ef51-705b-4bca-baf0-9a3209b9592b')
;

-- Aug 1, 2023, 11:58:24 AM VET
-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
INSERT INTO M_Product_Acct (AD_Client_ID,AD_Org_ID,C_AcctSchema_ID,Created,CreatedBy,IsActive,M_Product_ID,P_Asset_Acct,P_AverageCostVariance_Acct,P_Burden_Acct,P_COGS_Acct,P_CostAdjustment_Acct,P_CostOfProduction_Acct,P_Expense_Acct,P_FloorStock_Acct,P_InventoryClearing_Acct,P_InvoicePriceVariance_Acct,P_Labor_Acct,P_MethodChangeVariance_Acct,P_MixVariance_Acct,P_OutsideProcessing_Acct,P_Overhead_Acct,P_PurchasePriceVariance_Acct,P_RateVariance_Acct,P_Revenue_Acct,P_Scrap_Acct,P_TradeDiscountGrant_Acct,P_TradeDiscountRec_Acct,P_UsageVariance_Acct,P_WIP_Acct,Updated,UpdatedBy,UUID) VALUES (11,0,101,TO_DATE('2023-08-01 11:58:23','YYYY-MM-DD HH24:MI:SS'),100,'Y',1000000,231,50013,50008,233,298,50009,230,50002,299,282,50007,50003,50006,50010,50011,232,50005,229,50012,284,283,50004,50001,TO_DATE('2023-08-01 11:58:23','YYYY-MM-DD HH24:MI:SS'),100,'bde2946c-4ca0-486c-b1ab-e048948c2b02')
;

-- Aug 1, 2023, 11:58:24 AM VET
-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
INSERT INTO M_Product_Costing (AD_Client_ID,AD_Org_ID,C_AcctSchema_ID,Created,CreatedBy,IsActive,M_Product_ID,Updated,UpdatedBy,UUID) VALUES (11,0,101,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,'Y',1000000,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,'d660b996-afed-40f6-8d0d-0e1e169f3d83')
;

-- Aug 1, 2023, 11:58:24 AM VET
-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
INSERT INTO M_Cost (AD_Client_ID,AD_Org_ID,C_AcctSchema_ID,Created,CreatedBy,CumulatedAmt,CumulatedQty,CurrentCostPrice,CurrentQty,FutureCostPrice,IsActive,IsCostFrozen,M_AttributeSetInstance_ID,M_CostElement_ID,M_CostType_ID,M_Product_ID,M_Warehouse_ID,Updated,UpdatedBy,UUID) VALUES (11,0,101,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,0,0,0,0,0,'Y','N',0,100,100,1000000,0,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,'b0e975ce-5994-4ee4-bee2-a073237a2b4a')
;

-- Aug 1, 2023, 11:58:24 AM VET
-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
INSERT INTO M_Cost (AD_Client_ID,AD_Org_ID,C_AcctSchema_ID,Created,CreatedBy,CumulatedAmt,CumulatedQty,CurrentCostPrice,CurrentQty,FutureCostPrice,IsActive,IsCostFrozen,M_AttributeSetInstance_ID,M_CostElement_ID,M_CostType_ID,M_Product_ID,M_Warehouse_ID,Updated,UpdatedBy,UUID) VALUES (11,0,101,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,0,0,0,0,0,'Y','N',0,100,50000,1000000,0,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,'9d073a38-84c1-4058-ab48-981cf77c824e')
;

-- Aug 1, 2023, 11:58:25 AM VET
-- I forgot to set the DICTIONARY_ID_COMMENTS System Configurator
INSERT INTO AD_TreeNodePR (AD_Client_ID,AD_Org_ID,AD_Tree_ID,Created,CreatedBy,IsActive,Node_ID,Parent_ID,SeqNo,Updated,UpdatedBy,UUID) VALUES (11,0,102,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,'Y',1000000,0,999,TO_DATE('2023-08-01 11:58:24','YYYY-MM-DD HH24:MI:SS'),100,'3c6bb23e-6302-4bb6-ae89-6caf8d49bcee')
;
```

Best regards

**Reply by sdgathman** [reply]
Ok, I loaded the 2.4 million records into t_parts, and now I can populate i_product via query before running ImportProduct.   I recommend this approach for any dataset too big to use the webui conveniently.

I added a t_parts_id serial to facilitate batching (schlep a range of t_parts over to i_product) and debugging.

**Reply by sdgathman** [reply]
Just for fun, I kicked off import of all 2 million products.  It looks like the entire thing is loaded into a single ginormous transaction.  :-(

Definitely need an easy way to split it up.  I'm thinking a 3rd state for the isimported flag on i_product.

**Reply by piracio** [reply]
Hello, if you like to speed the process let's schedule a meeting when you have time, I can take a look at database level what is going on with the import, usually commits are done/should be done each X records, stream load methods should be used to reduce the memory usage, GC/cpu usage. 

At transport level optimize the http layer to utilize chunk sizes of 8K as peer standard and perhaps we should change the approach split the load into X numbers of records, heavy loads or large databases are my specialty. If you have time, we can use the adempiere demo server for that.

**Reply by sdgathman** [reply]
For now, I am modifying my sql script to copy from t_parts to i_product in small batches.  Then, I can run the import, and records finished are flagged with isimported.  Lather, rinse, repeat.

I will ask the client if the parts list can be shared for development on demo server.  Otherwise, we can use my demo server.

**Reply by sdgathman** [reply]
@piracio I set ad_client_id to 0 for all but 46 records in i_product (so that only 46 products are imported instead of 2 million).
It has been 45 minutes.  I can see which step it is on with  "select pid,query from pg_stat_activity;" 

I intended to work up to larger batches and time the imports.  But this is ridiculous for just 46 records.  I suspect part of the problem is that i_product needs some indexes.  I think an index on ad_client_id may be important when using ad_client_id to select records to be imported.  Or else, delete records from i_product and add them back in batches.

Update: it is done - took about a minute per product.  The webui hasn't noticed yet.

**Reply by sdgathman** [reply]
@piracio First PR for the low handing fruit: https://github.com/adempiere/adempiere/pull/4139

**Comment by piracio** [comment]
I did not play with the imports of the Adempiere, I did some migration indexes improvements.

I am not functional if someone is able to replicate the issue that you are having to take a look will be great to fix this missing indexes and/or the logic of the Imports.

**Reply by sdgathman** [reply]
Doing the import of 50 products with only 107 records in i_product.  Still taking over a minute per product.  Is there a tool to log the queries being done, with the time?  I vaguely recall something.  That would indicate which queries to optimize.

**Reply by sdgathman** [reply]
This query (error check) is taking quite some time:
`UPDATE I_Product SET I_IsImported='E', I_ErrorMsg=I_ErrorMsg||'ERR=UPC 
not unique,' WHERE I_IsImported<>'Y' AND UPC IN (SELECT UPC FROM I_Product ii WH
ERE I_Product.AD_Client_ID=ii.AD_Client_ID GROUP BY UPC HAVING COUNT(*) > 1) AND
 AD_Client_ID=1000001`

**Comment by yamelsenih** [comment]
Very bad query, I think that is better if you change the `IN` for a `EXISTS` clause

**Reply by sdgathman** [reply]
So now I have to setup a debug cycle where I can compile the java class (where the query is hardwired).  The batched import is important, but only with a reasonable time per record.

**Reply by sdgathman** [reply]
So this is a test query to just list the parts that would get an error (should be 0).  There are 107 records to check.  For debugging purposes, it checks all records, even those already imported.
```
SELECT value,I_IsImported from I_Product
WHERE /* I_IsImported<>'Y' AND  */
EXISTS(SELECT UPC from I_Product ii WHERE I_Product.AD_Client_ID=ii.AD_Client_ID AND ii.UPC = I_Product.UPC AND ii.I_Product_id != I_Product.I_Product_id)
/* UPC IN (SELECT UPC FROM I_Product ii WHERE I_Product.AD_Client_ID=ii.AD_Client_ID GROUP BY UPC HAVING COUNT(*) > 1) */
AND AD_Client_ID=1000001;
```
My first change to EXISTS did not improve anything.

**Reply by sdgathman** [reply]
On the postscript side, it is all CPU (unsurprisingly with only 107 records).   I interrupted after 6.5 mins.  This should be instant with that number of records.

**Reply by sdgathman** [reply]
@yamelsenih see PR above.  IN/EXISTS by itself do not change query plan: https://www.percona.com/blog/sql-optimizations-in-postgresql-in-vs-exists-vs-any-all-vs-join/ 
The issue with that (and similar) queries was self reference for AD_Client_ID making time proportional to N².  I changed to a constant (which is already known).  There are more such problems I am sure.

**Reply by yamelsenih** [reply]
Hello @sdgathman si think that is better change a IN with a query for a EXISTS, the IN clause should be used only for static values.

```
IN('CO', 'CL')
```
But a query inside a `IN` is very lazy for query
